src/const-generics.md

```diff
@@ -0,0 +1,174 @@
+parameter.
+
+r[items.generics.const.type-ambiguity]
+When there is ambiguity if a generic argument could be resolved as either a
+type or const argument, it is always resolved as a type. Placing the argument
+in a block expression can force it to be interpreted as a const argument.
+
+<!-- TODO: Rewrite the paragraph above to be in terms of namespaces, once
+    namespaces are introduced, and it is clear which namespace each parameter
+    lives in. -->
+
+```rust,compile_fail
+type N = u32;
+struct Foo<const N: usize>;
+// The following is an error, because `N` is interpreted as the type alias `N`.
+fn foo<const N: usize>() -> Foo<N> { todo!() } // ERROR
+// Can be fixed by wrapping in braces to force it to be interpreted as the `N`
+// const parameter:
+fn bar<const N: usize>() -> Foo<{ N }> { todo!() } // ok
+```
+
+r[items.generics.const.exhaustiveness]
+When resolving a trait bound obligation, the exhaustiveness of all
+implementations of const parameters is not considered when determining if the
+bound is satisfied. For example, in the following, even though all possible
+const values for the `bool` type are implemented, it is still an error that
+the trait bound is not satisfied:
+
+```rust,compile_fail
+struct Foo<const B: bool>;
+trait Bar {}
+impl Bar for Foo<true> {}
+impl Bar for Foo<false> {}
+
+fn needs_bar(_: impl Bar) {}
+fn generic<const B: bool>() {
+    let v = Foo::<B>;
+    needs_bar(v); // ERROR: trait bound `Foo<B>: Bar` is not satisfied
+}
+```
+
+r[const-generics.kinds]
+There are three kinds of arguments to a const parameter:
+1. Standalone const parameters
+2. Inferred consts
+3. Arbitrary concrete expressions 
+
+r[const-generics.standalone]
+## Standalone const parameters
+
+A const parameter can only be used in a const argument if it is a standalone usage.
+The argument must be *only* a usage of a const parameter and can be wrapped
+in at most one level of braces. That is, they cannot be combined with other
+expressions.
+```rust
+// Examples of standalone uses of const parameters
+
+fn foo<const N: usize>() {
+    let a: [u8; N] = [10; N];
+    let b: [u8; { N }] = a;
+    foo::<N>();
+    foo::<{ N }>();
+}
+```
+
+Here `a` has type `[u8; N]`, an array with a length of `N`, referring to `foo`'s const parameter.
+
+```rust,compile_fail
+// Examples of non-standalone uses of const parameters
+
+fn foo<const N: usize>() {
+    let a: [u8; {{ N }}] = [10; (N)];
+    foo::<{{ N }}>();
+    foo::<(N)>();
+}
+```
+
+r[const-generics.inferred]
+## Inferred consts
+
+r[const-generics.inferred.syntax]
+```grammar,types
+@root InferredConst -> `_`
+```
+
+The inferred const asks the compiler to infer the const if possible based on
+the surrounding information available.
+
+It cannot be used in item signatures.
+
+It is often used in repeat expressions:
+```rust
+fn make_array() -> [u32; 2] {
+    [Default::default(); _]
+}
+```
+
+r[const-generics.concrete-expr]
+## Concrete expressions
```

Comment:
Something I left out here is that the "concrete expressions" are not required to be quite so concrete for repeat expr counts as a backwards compatibility FCW. Does the reference intend to document the *actual state* of the compiler, or can I take liberties in documenting things as if FCWs are already hard errors?